### PR TITLE
Wire auditing into refinement chain

### LINKED ISSUE #405: Wire auditing into refinement chain as first-class lifecycle step
## Summary

Auditing is currently siloed: `codebase-deslop-audit` runs as a one-off epic, and the `deep` persona (#402) is opt-in per-PR review. Neither is wired into the dev-loop lifecycle.

Auditing should be a first-class capability usable in both phases:
- **Refinement:** run a scoped deslop pass before planning to surface slop risks
- **Review:** the `deep` persona catches regressions per PR

## Why this phase exists now

Refinement currently goes straight from issue/context gathering into fan-out planning. That leaves a gap: known structural slop risks, oversized-file risk, duplication risk, and adjacent cleanup opportunities are discovered too late, after scope and AC/DoD are already hardening.

This issue closes only the refinement-side lifecycle gap. It makes a bounded, opt-in audit pass a first-class pre-step before refiner fan-out, then teaches the refiner to consume that audit output explicitly. Review-side structural regression coverage remains the separate #402 slice.

### In scope

- document an opt-in bounded refinement chain shape:
  - `audit -> refine -> implement`
- define where that audit step is documented and how it is passed into refinement
- add one reusable bounded audit prompt plus one deterministic audit helper contract
- extend the `refiner` contract so audit findings become first-class planning inputs instead of side notes
- define deterministic audit output/artifact expectations for local phase refinement and GitHub issue refinement
- add contract and CLI tests that prove the audit artifact can flow into the refinement chain without inventing a new orchestration system

### Explicit non-goals

- Not replacing `codebase-deslop-audit` (full-repo)
- Not requiring audit in every chain (opt-in)

### Review-side dependency / intentionally unchanged surface

- `skills/copilot-pr-followup/SKILL.md` should stay unchanged in this issue.
- Reason: the review side already resolves pre-approval gate angles from config; #402 remains the separate review-side implementation for the `deep` structural-audit persona.
- This issue should not broaden into PR-gate behavior, reviewer-loop semantics, or new mandatory review gates.

### `AGENTS.md`

Update **Standard refinement chain pattern** so the canonical repo-level wording becomes:
- refinement may start with an opt-in bounded audit pre-step when the issue/phase has likely slop risk, structural drift risk, or adjacent cleanup risk
- when used, the sequence is explicitly `audit -> refine -> implement`
- the audit must be bounded to named files/areas; it must not silently widen into a full-repo scan
- audit output must be prioritized and passed into refiner fan-out/fan-in as an input artifact, not pasted as an unstructured aside
- the audit step must state what the current phase will **not** rewrite or broaden

### `skills/local-implementation/SKILL.md`

Add the operational refinement-side workflow seam that `AGENTS.md` points at:
- before variant fan-out, optionally run one bounded audit when the active phase would benefit from slop-risk discovery
- store the artifact under a deterministic phase-scoped path such as:
  - `tmp/phases/phase-x/audit/refinement-audit-summary.json`
- pass a concise audit summary into every refiner briefing
- require the merged plan/review artifacts to preserve prioritized findings, highest-value follow-up candidates, and an explicit statement of what the phase will not rewrite or broaden
- keep this opt-in; no default mandatory audit run

### `skills/docs/issue-intake-procedure.md`

Update **Phase 3 — Async issue refinement** so GitHub-first issue refinement can use the same bounded audit seam before issue-body hardening:
- when issue refinement would benefit from structural/slop discovery and the likely code/doc surface is knowable, run the bounded audit first
- keep issue refinement separate from local phase refinement, but let both consume the same audit artifact shape
- require the issue-refinement write-up to translate audit findings into scope, AC/DoD, risks, and explicit non-goals without silently broadening the issue

### `agents/refiner.agent.md`

Extend the refiner contract so that when an audit artifact is provided, the refiner must:
- consume it explicitly
- surface prioritized findings and highest-value follow-up candidates
- convert each meaningful audit finding into exactly one of:
  - current-phase scope/acceptance-criteria change
  - definition-of-done expectation
  - explicit non-goal / defer
  - risk/watchpoint
- state explicitly what the current phase will **not** rewrite or broaden
- avoid inventing audit findings when no audit artifact exists

### `.pi/dev-loop/defaults.yaml`

Make two prompt-level changes:

1. extend `personas.refiner.prompt` with an audit-consumption addendum, for example:
   - when a bounded audit summary is supplied, add an `Audit inputs` subsection
   - classify each prioritized finding as scope, AC, DoD, risk/watchpoint, or deferred follow-up
   - include the highest-value follow-up candidates
   - include an explicit `Will not rewrite/broaden in this phase` statement
   - do not fabricate audit evidence when none was provided

2. add a reusable non-gate persona prompt, for example `personas.audit`, that borrows the deletion-first posture from `codebase-deslop-audit` but stays phase-bounded:
   - audit only the named files/areas
   - prefer delete / merge / trim / defer framing
   - produce prioritized findings, not a whole-repo essay
   - always include a bounded-scope statement and a “not rewriting in this phase” statement

### `scripts/loop/run-refinement-audit.mjs` (new)

Add one deterministic helper for bounded refinement audits.

**Purpose**
- provide one reusable machine-readable audit artifact for refinement chains
- complement the `personas.audit` prompt with mechanical heuristics
- stay bounded to explicit scope; do not replace full-repo audit work

**CLI contract**
- required: one of
  - `--paths <comma-separated path list>`
  - `--paths-file <file>`
- optional:
  - `--root <path>` (defaults to repo root)
  - `--max-lines <n>` (default `1000`)
  - `--duplicate-window-lines <n>` (default `4`)
  - `--branch-threshold <n>` (default `25` control-flow tokens)
  - `--thin-wrapper-max-lines <n>` (default `40`)
  - `--output <path>` writes the same success JSON emitted on stdout

**Scope rules**
- no `--paths*` input => deterministic argument error; never silently scan the whole repo
- directory inputs expand through tracked files only (`git ls-files -- <path>...`)
- duplicate path inputs are de-duped
- binary/unreadable files are skipped with machine-readable notes, not fatal unless nothing auditable remains

**Heuristic findings for v1**
- `oversized_file`
  - line count exceeds `--max-lines`
- `duplicate_block_candidate`
  - normalized repeated text block of at least `--duplicate-window-lines` appears more than once within the bounded scope
- `branching_hotspot`
  - control-flow token count exceeds `--branch-threshold`
- `thin_wrapper_candidate`
  - small file dominated by re-exports / passthrough-only glue and little or no distinct logic

**Success output shape**
```json
{
  "ok": true,
  "repoRoot": "/repo",
  "paths": ["AGENTS.md", "agents/refiner.agent.md"],
  "auditedFiles": [
    {
      "path": "agents/refiner.agent.md",
      "lineCount": 84,
      "branchTokenCount": 0,
      "duplicateBlockMatches": 0,
      "thinWrapperCandidate": false,
      "skipped": false
    }
  ],
  "findings": [
    {
      "id": "oversized_file",
      "priority": "high",
      "path": "skills/local-implementation/SKILL.md",
      "summary": "File exceeds the bounded audit line threshold.",
      "evidence": { "lineCount": 520, "threshold": 400 }
    }
  ],
  "highestValueFollowUpCandidates": [
    {
      "path": "skills/local-implementation/SKILL.md",
      "reason": "Carries both workflow contract text and audit guidance; trim or split only if this phase explicitly chooses to."
    }
  ],
  "scopeBoundary": {
    "mode": "bounded_paths_only",
    "fullRepoScan": false
  }
}
```

**Failure behavior**
- malformed arguments, empty/blank paths, invalid thresholds, or zero auditable files after expansion emit JSON on stderr and exit non-zero
- finding presence is **not** a process failure; findings return exit `0`

**Integration expectation**
- this remains repo-local workflow glue under `scripts/loop/`
- local callers use the repo-local script path directly
- do not create a new published package surface just for this helper unless multiple runtime callers genuinely need shared parsing logic later

### `scripts/README.md`

Document the new helper with:
- usage
- success output shape
- failure behavior
- the bounded-scope rule (`--paths*` required; no silent whole-repo scan)
- the fact that findings are planning inputs, not auto-authorized rewrites

### `test/loop/run-refinement-audit.test.mjs` (new)

Add deterministic CLI/helper tests for:
- missing scope args fail closed
- directory expansion uses tracked files only
- oversized-file finding is emitted
- duplicate-block candidate is emitted
- branching-hotspot finding is emitted
- thin-wrapper candidate is emitted
- binary/unreadable files are skipped deterministically
- `--output` writes the same JSON returned on stdout
- clean bounded scopes return `ok: true` with empty findings

### `test/loop/refinement-audit-chain-contract.test.mjs` (new)

Add a narrow end-to-end contract test for the refinement handoff boundary:
- create a bounded fixture scope
- run `run-refinement-audit.mjs`
- assert the emitted JSON contains the fields the refiner prompt/contract expects:
  - `findings`
  - `highestValueFollowUpCandidates`
  - `scopeBoundary`
- verify deterministic field presence/order assumptions used by the documented refinement chain

This keeps “tested end-to-end in a refinement chain” honest without inventing a new live-agent orchestrator test harness.

### `test/contracts/planning-doc-contracts.test.mjs`

Extend planning contract coverage so tests fail if the docs/prompts drift from this plan:
- `AGENTS.md` must mention the opt-in `audit -> refine -> implement` chain and bounded audit rule
- `agents/refiner.agent.md` must mention consuming audit findings and stating what the phase will not rewrite/broaden
- `.pi/dev-loop/defaults.yaml` must include the refiner audit-consumption prompt text and the reusable audit persona prompt
- `skills/local-implementation/SKILL.md` and `skills/docs/issue-intake-procedure.md` must mention the audit artifact handoff before refinement hardening

### `skills/copilot-pr-followup/SKILL.md`

No change in this issue.

Reason:
- this skill already resolves review angles from config
- review-side structural auditing belongs to #402
- changing PR follow-up behavior here would broaden the issue beyond the accepted phase boundary

## Refiner prompt extension

Add wording equivalent to:

> When a bounded audit artifact is provided, treat it as a first-class planning input. Summarize the audited scope, list prioritized findings, include the highest-value follow-up candidates, and explicitly state what the current phase will not rewrite or broaden. For each prioritized finding, classify it as one of: current-phase scope/AC, DoD, explicit non-goal/defer, or risk/watchpoint. Do not invent audit findings when no audit artifact was provided.

## Complete acceptance criteria

1. `audit` step documented in refinement chain pattern (AGENTS.md)
   - `AGENTS.md` documents the opt-in bounded `audit -> refine -> implement` sequence
   - the audit step is explicitly bounded to named files/areas
   - the handoff into refiner fan-out/fan-in is explicit

2. `refiner` agent prompt updated to consume audit findings
   - both `agents/refiner.agent.md` and `.pi/dev-loop/defaults.yaml` explain how audit findings are consumed
   - the refiner must convert findings into scope, AC, DoD, non-goal/defer, or risk/watchpoint outputs
   - the refiner must explicitly state what the current phase will not rewrite or broaden

3. At least one concrete audit script/prompt reusable in chains
   - `personas.audit` provides the reusable prompt
   - `scripts/loop/run-refinement-audit.mjs` provides the deterministic helper
   - both stay bounded and opt-in

4. Tested end-to-end in a refinement chain
   - CLI tests cover the audit helper contract
   - planning contract tests cover the documented handoff into refinement
   - one narrow chain-contract test proves the audit JSON shape matches the refiner-consumption contract

## Proposed DoD

- The canonical refinement docs describe the opt-in bounded `audit -> refine -> implement` chain and preserve the opt-in/non-broadening boundary.
- A reusable bounded audit surface exists as both a prompt (`personas.audit`) and a deterministic helper (`scripts/loop/run-refinement-audit.mjs`).
- The refiner contract and default refiner prompt both explain how audit findings change scope, AC/DoD, non-goals/defer, and risks/watchpoints.
- Local phase refinement and GitHub issue refinement both document where the audit artifact is created/consumed before issue/phase hardening.
- Tests cover doc-contract drift, audit-helper CLI behavior, and the audit-artifact handoff shape into refinement.
- The implementation leaves review-side behavior to #402 and does not replace `codebase-deslop-audit` or make audit mandatory.

## Risks and watchpoints

- **False precision from heuristics:** the audit helper should surface candidates, not pretend to prove architectural truth. Keep the refiner responsible for turning findings into plan decisions.
- **Scope inflation risk:** if `--paths` is omitted or too broad, the audit can turn into a stealth full-repo exercise. Fail closed on missing scope and document that boundedness is mandatory.
- **Threshold churn:** line-count / duplicate-window / branch-threshold defaults must be documented and tested so follow-up tweaks are deliberate.
- **Dependency sequencing:** review-side parity still depends on #402 landing separately. This issue should document that dependency rather than quietly absorbing it.

## Unresolved questions

None requiring RFC treatment.

## Durable decisions to preserve in the issue/phase doc

- Audit remains **opt-in**, never an always-on precondition.
- Audit scope is **bounded to explicit paths**; no silent full-repo scan.
- Audit findings are **planning inputs**, not automatic rewrite authorization.
- Review-side structural auditing remains the separate #402 track because `skills/copilot-pr-followup/SKILL.md` already consumes config-resolved pre-approval angles.
- Keep the helper as thin repo-local workflow glue under `scripts/loop/` unless later reuse justifies extraction.

## Tests to write first

1. `test/loop/run-refinement-audit.test.mjs`
   - start with malformed-argument coverage
   - then add one fixture per finding type
2. `test/loop/refinement-audit-chain-contract.test.mjs`
   - lock the JSON handoff shape before any implementation refactoring
3. `test/contracts/planning-doc-contracts.test.mjs`
   - lock doc/prompt wording so the lifecycle contract does not regress

## Validation steps

- `node --test test/loop/run-refinement-audit.test.mjs test/loop/refinement-audit-chain-contract.test.mjs test/contracts/planning-doc-contracts.test.mjs`
- `npm test`
- targeted read-through of:
  - `AGENTS.md`
  - `skills/local-implementation/SKILL.md`
  - `skills/docs/issue-intake-procedure.md`
  - `agents/refiner.agent.md`
  - `.pi/dev-loop/defaults.yaml`
  - `scripts/README.md`
- manual sanity check that the issue body and docs all preserve the same boundary:
  - bounded audit
  - opt-in only
  - not a full-repo replacement
  - review-side remains #402

## RFC escalation needed through the coordinator

None. This plan stays inside existing doc/prompt/script surfaces, avoids a new config family, and does not change public workflow routing or PR-gate semantics.

## AC/DoD/Non-goal coverage matrix

| Item | Type (AC/DoD/Non-goal) | Status (Met/Partial/Unmet/Unverified) | Evidence | Notes |
|---|---|---|---|---|
| `audit` step documented in refinement chain pattern (AGENTS.md) | AC | Met | `Exact changes per file` -> `AGENTS.md`; `Complete acceptance criteria` item 1 | Opt-in bounded `audit -> refine -> implement` chain is specified. |
| `refiner` agent prompt updated to consume audit findings | AC | Met | `Exact changes per file` -> `agents/refiner.agent.md`, `.pi/dev-loop/defaults.yaml`; `Refiner prompt extension`; `Complete acceptance criteria` item 2 | Refiner must classify audit findings into scope/AC/DoD/non-goal/risk outputs. |
| At least one concrete audit script/prompt reusable in chains | AC | Met | `Exact changes per file` -> `.pi/dev-loop/defaults.yaml`, `scripts/loop/run-refinement-audit.mjs`, `scripts/README.md`; `Complete acceptance criteria` item 3 | Plan includes both a reusable prompt and deterministic helper. |
| Tested end-to-end in a refinement chain | AC | Met | `Exact changes per file` -> `test/loop/refinement-audit-chain-contract.test.mjs`, `test/contracts/planning-doc-contracts.test.mjs`; `Tests to write first`; `Validation steps`; `Complete acceptance criteria` item 4 | End-to-end here is the bounded audit-artifact-to-refiner contract, not a new live-agent harness. |
| The canonical refinement docs describe the opt-in bounded `audit -> refine -> implement` chain and preserve the opt-in/non-broadening boundary. | DoD | Met | `Exact changes per file` -> `AGENTS.md`, `skills/local-implementation/SKILL.md`, `skills/docs/issue-intake-procedure.md`; `Validation steps` | Covers both local phase refinement and GitHub issue refinement docs. |
| A reusable bounded audit surface exists as both a prompt (`personas.audit`) and a deterministic helper (`scripts/loop/run-refinement-audit.mjs`). | DoD | Met | `Exact changes per file` -> `.pi/dev-loop/defaults.yaml`, `scripts/loop/run-refinement-audit.mjs`, `scripts/README.md` | Keeps audit reusable without making it mandatory. |
| The refiner contract and default refiner prompt both explain how audit findings change scope, AC/DoD, non-goals/defer, and risks/watchpoints. | DoD | Met | `Exact changes per file` -> `agents/refiner.agent.md`, `.pi/dev-loop/defaults.yaml`; `Refiner prompt extension` | Preserves the audit-to-plan translation rules. |
| Local phase refinement and GitHub issue refinement both document where the audit artifact is created/consumed before issue/phase hardening. | DoD | Met | `Exact changes per file` -> `skills/local-implementation/SKILL.md`, `skills/docs/issue-intake-procedure.md` | Avoids documenting the seam in only one workflow. |
| Tests cover doc-contract drift, audit-helper CLI behavior, and the audit-artifact handoff shape into refinement. | DoD | Met | `Exact changes per file` -> `test/contracts/planning-doc-contracts.test.mjs`, `test/loop/run-refinement-audit.test.mjs`, `test/loop/refinement-audit-chain-contract.test.mjs`; `Tests to write first`; `Validation steps` | Explicitly includes negative-case CLI tests first. |
| The implementation leaves review-side behavior to #402 and does not replace `codebase-deslop-audit` or make audit mandatory. | DoD | Met | `Review-side dependency / intentionally unchanged surface`; `Durable decisions to preserve in the issue/phase doc`; `Explicit non-goals` | Keeps this phase refinement-bounded. |
| Not replacing `codebase-deslop-audit` (full-repo) | Non-goal | Met | `Explicit non-goals`; `Review-side dependency / intentionally unchanged surface`; `Durable decisions to preserve in the issue/phase doc` | Bounded audit helper complements, not replaces, the full-repo skill. |
| Not requiring audit in every chain (opt-in) | Non-goal | Met | `Explicit non-goals`; `Exact changes per file` -> `AGENTS.md`, `skills/local-implementation/SKILL.md`; `Durable decisions to preserve in the issue/phase doc` | Audit remains available but not mandatory. |

### PR DESCRIPTION
**Pull request by mfittko**
## Summary
- add a deterministic bounded refinement audit helper at `scripts/loop/run-refinement-audit.mjs`
- wire the opt-in `audit -> refine -> implement` refinement seam into repo docs and refiner prompts
- add CLI, chain-contract, and planning-doc contract coverage for the audit handoff

## Scope / context
Implements #405 by making refinement-side auditing a first-class planning input without changing review-side PR follow-up behavior.

## Acceptance criteria
- `AGENTS.md` documents the opt-in bounded `audit -> refine -> implement` sequence.
- `agents/refiner.agent.md` and `.pi/dev-loop/defaults.yaml` explain how audit findings are consumed.
- `.pi/dev-loop/defaults.yaml` includes a reusable `personas.audit` prompt.
- `skills/local-implementation/SKILL.md` and `skills/docs/issue-intake-procedure.md` document where the audit artifact is created/consumed before refinement hardening.
- `scripts/loop/run-refinement-audit.mjs` implements the bounded audit helper contract.
- Tests cover CLI behavior, planning-doc drift, and the audit-artifact handoff shape into refinement.

## Definition of done
- local `npm run verify` passes
- the new audit helper emits deterministic success/error JSON and never silently scans the whole repo
- docs/prompts preserve the opt-in, bounded, non-broadening posture
- review-side behavior remains unchanged and stays with #402

## Non-goals
- replacing `codebase-deslop-audit`
- making refinement audit mandatory
- changing `skills/copilot-pr-followup/SKILL.md` or review-side gate behavior

## Validation
- `node --test test/loop/run-refinement-audit.test.mjs test/loop/refinement-audit-chain-contract.test.mjs test/contracts/planning-doc-contracts.test.mjs`
- `npm run verify`

Closes #405

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `8a2494d180e3bf13bd0f0b30ff3b890de7d7c0e2`
**Verdict:** clean

**Findings summary:** Draft gate clean: scope, coverage, and correctness review passed; npm run verify passed locally and on CI; changes stay within issue #405.

**Next action:** Mark PR ready for review and request Copilot review.

File: scripts/loop/run-refinement-audit.mjs [hunk 1]
```diff
@@ -0,0 +1,593 @@
...
+}
+
+async function loadConfiguredPaths(options, cwd) {
+  if (options.paths !== undefined) {
+    return assertNoBlankPaths(splitConfiguredPaths(options.paths));
+  }
+
+  const filePath = path.resolve(cwd, options.pathsFile);
+  const raw = await readFile(filePath, "utf8");
+  const entries = raw.split(/\r?\n/u);
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/loop/run-refinement-audit.mjs:191]: Fixed in 57f094c: unreadable --paths-file reads now fail closed as CLI parse errors with usage, and the regression test locks the stderr JSON shape.

File: test/loop/run-refinement-audit.test.mjs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    const result = await runAudit(["--root", repoRoot, "--paths-file", pathsFile]);
+    assert.equal(result.code, 0, result.stderr);
+
+    const parsed = JSON.parse(result.stdout.trim());
+    assert.deepEqual(parsed.paths, ["a.md", "b.md"]);
+    assert.deepEqual(parsed.auditedFiles.map((entry) => entry.path), ["a.md", "b.md"]);
+  } finally {
+    await cleanupRepo(repoRoot);
+  }
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply test/loop/run-refinement-audit.test.mjs:336]: Fixed in 57f094c: unreadable --paths-file reads now fail closed as CLI parse errors with usage, and the regression test locks the stderr JSON shape.

File: scripts/loop/run-refinement-audit.mjs [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+      duplicateCountsByPath.set(value.path, (duplicateCountsByPath.get(value.path) ?? 0) + 1);
+    }
+  }
+
+  return duplicateCountsByPath;
+}
+
+function isCommentLine(line) {
+  return line.startsWith("#") || line.startsWith("//") || line.startsWith("/*") || line.startsWith("*") || line.startsWith("*/");
+}
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/loop/run-refinement-audit.mjs:323]: Fixed in 572626c: generator-style  lines are no longer treated as comments for thin-wrapper detection, and repo-root path requests now fail closed so  cannot expand to a whole-repo scan.

File: scripts/loop/run-refinement-audit.mjs [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+    : path.resolve(repoRoot, requestedPath);
+  const relativePath = path.relative(repoRoot, resolvedPath);
+
+  if (relativePath.length === 0) {
+    return ".";
+  }
+
+  if (relativePath.startsWith(`..${path.sep}`) || relativePath === ".." || path.isAbsolute(relativePath)) {
+    throw parseError(`Path is outside the repo root: ${requestedPath}`);
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/loop/run-refinement-audit.mjs:223]: Fixed in 572626c: generator-style  lines are no longer treated as comments for thin-wrapper detection, and repo-root path requests now fail closed so  cannot expand to a whole-repo scan.

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `572626c6d6e1232050459afeb5d529802433307b`
**Verdict:** clean

**Findings summary:** Pre-approval gate clean: DRY/KISS/YAGNI/SRP/SoC review passed; npm run verify passed locally and on CI; current-head Copilot review is clean with no unresolved threads.

**Next action:** Ready for final approval and merge.
